Add developer docs for unique attributes. (#2289)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/api/schemas/index.md

Okta has also extended [JSON Schema Draft 4](https://tools.ietf.org/html/draft-zyp-json-schema-04) with the following keywords:

| Property                                                        | Description                                                       | DataType                                                                  | Nullable | Unique | Readonly |
| :-------------------------------------------------------------- | :--------------------------------------------------------------- | :------------------------------------------------------------------------ | :------- | :----- | :------- |
| required                                                        | Determines whether the property is required                       | Boolean                                                                   | FALSE    | FALSE  | FALSE    |
| permissions                                                     | Access control permissions for the property                       | Array of [Schema property permission](#schema-property-permission-object) | FALSE    | FALSE  | FALSE    |

> | [unique](#unique-attribute) <ApiLifecycle access="ea" />        | Determines whether the property enforces uniqueness constraint    | Boolean                                                                   | FALSE    | FALSE  | FALSE    |

Author: For ease of review:
![Screen Shot 2021-06-30 at 12 24 23 PM](https://user-images.githubusercontent.com/12566503/124019861-6fe0dd80-d99e-11eb-87c0-0f38dbb678d8.png)

> | [unique](#unique-attribute) <ApiLifecycle access="ea" />        | Determines whether the property enforces uniqueness constraint    | Boolean                                                                   | FALSE    | FALSE  | FALSE    |

Reviewer: "Enforces uniqueness constraint" feels self-referential.  How about: "Determines whether property values must be unique"
(I don't know if it's necessary to include the caveat that empty values don't count.)

Author: > I don't know if it's necessary to include the caveat that empty values don't count.

Specified that at L2480.

> | [unique](#unique-attribute) <ApiLifecycle access="ea" />        | Determines whether property values must be unique                 | Boolean                                                                   | FALSE    | FALSE  | FALSE    |

Author: No reason on the current ordering. I just added the `unique`  attribute to the last row as it's new. Now that you brought it up I have the same feeling that `required` and `unique` share more similarity. Moved `unique` up.

Okta has also extended [JSON Schema Draft 4](https://tools.ietf.org/html/draft-zyp-json-schema-04) with the following keywords:

| Property                                                   | Description                                                                      | DataType                    | Nullable | Unique | Readonly |
| :--------------------------------------------------------- | :------------------------------------------------------------------------------- | :-------------------------- | :------- | :----- | :------- |
| required                                                   | Determines whether the property is required                                      | Boolean                     | FALSE    | FALSE  | FALSE    |
| scope                                                      | Determines whether a group attribute can be set at the individual or group level | `SELF`, `NONE`              | FALSE    | FALSE  | TRUE     |

> | [unique](#unique-attribute) <ApiLifecycle access="ea" />   | Determines whether the property enforces uniqueness constraint                   | Boolean                     | FALSE    | FALSE  | FALSE    |

Author: For ease of review:
![Screen Shot 2021-06-30 at 12 24 35 PM](https://user-images.githubusercontent.com/12566503/124019915-7e2ef980-d99e-11eb-8a08-ad96d73c710d.png)

> | [unique](#unique-attribute) <ApiLifecycle access="ea" />   | Determines whether the property enforces uniqueness constraint                   | Boolean                     | FALSE    | FALSE  | FALSE    |

Reviewer: ibid

| `number`      | [JSON Number](https://tools.ietf.org/html/rfc7159#section-6) with double-precision 64-bit IEEE 754 floating point number constraint | `minimum` and `maximum`     |
| `integer`     | [JSON Number](https://tools.ietf.org/html/rfc7159#section-6) with 32-bit signed two's complement integer constraint           | `minimum` and `maximum`     |
| `array`       | [JSON Array](https://tools.ietf.org/html/rfc7159#section-5)                                                                         |                             |

## Unique attribute

<ApiLifecycle access="ea" />

> You can enforce uniqueness for custom properties in the Okta user profile or the Okta group profile, such as employee identification number. You can declare a maximum of 5 unique properties for each user type, and 5 unique properties in the Okta group profile. For unique properties in Okta user profile, the limit of 5 unique properties is applied per-type. You do not need to select the same set of properties for each user type. For example, the 5 custom properties you declare unique for user profile A do not need to match what you declared for user profiles B, C, or D.

Reviewer 2: ```suggestion
You can enforce uniqueness for custom properties in the Okta user profile or the Okta group profile, such as an employee identification number. You can declare a maximum of five unique properties for each user type and five unique properties in the Okta group profile. For unique properties in the Okta user profile, the limit of five unique properties is applied per type. You don't need to select the same set of properties for each user type. For example, the five custom properties that you declare unique for user profile A don't need to match what you declared for user profiles B, C, or D.
```

Reviewer: The third sentence is redundant; the previous sentence already says the limit for user properties is "for each user type".  Indeed, I suggest removing all the remaining sentences and condensing them to:
Different user types can have the same or different unique properties (up to the limit of five per type).

I was going to continue with details about interactions between types but I see those have their own paragraphs below, which is probably better.

Author: Updated.

> Unique properties in the Okta user profile share a single namespace across all [user types](/docs/reference/api/user-types) in an org. If user types A and B both contain a property ice cream and you identify it as unique in both profiles, then if user type A has the value chocolate, no other users of type A or B (or any other user type with ice cream declared unique) can have that value.

Reviewer 2: ```suggestion
Unique properties in the Okta user profile share a single namespace across all [user types](/docs/reference/api/user-types) in an org. If user types A and B both contain the property **ice cream** and you identify it as unique in both profiles, then if user type A has the value **chocolate**, no other users of type A or B (or any other user type with **ice cream** declared unique) can have that value.
```

Reviewer: I agree with @susanharper-okta to include highlighting, but here she uses star-star to highlight the property name and value, while in the example below she suggests using back-quotes.  Which format is correct?

I also agree with her other minor edit ("the").

Author: Changed to ` as Susan suggested in another comment.

> Properties that are not unique are not tracked for uniqueness. If the property candy is unique in type E and not unique in type F, and a user of type E has the value caramel for the property, then no other users of type E can have the value caramel for the property, but any number of users of type F can have the value caramel. Although candy is unique in E, it is not unique in F, so the value for the property in users of type F does not matter.

Reviewer 2: ```suggestion
Properties that aren't unique aren't tracked for uniqueness. If the property `candy` is unique in type E and not unique in type F, and a user of type E has the value `caramel` for the property, then no other users of type E can have the value `caramel` for the property. But any number of users of type F can have the value `caramel`. Although `candy` is unique in E, it isn't unique in F, so the value for the property in users of type F doesn't matter.
```

Reviewer: I agree with most of Susan's suggestions, but I don't like starting a sentence with "But" so I would leave that as "comma but" as in the original.  And again, which highlighting form is correct?

Reviewer 2: missed one - should just be `ice cream`

Reviewer 2: i updated in the previous comment - my apologies...thought I caught them all.

Reviewer 2: > I agree with most of Susan's suggestions, but I don't like starting a sentence with "But" so I would leave that as "comma but" as in the original. And again, which highlighting form is correct?

@donwoods-okta - I was trying to split up the sentence as it is incredibly long. Have a suggestion on how we can split that sentence up into two? It's a lot to take in.

> If you attempt to create or update a user with a duplicate value for a custom user property with a uniqueness restriction, the user creation or update operation will fail. The user won't be created or updated until you enter a unique value. Similarly, you would need to provide a unique value for a unique custom group property in order to create or update a group.

Reviewer 2: ```suggestion
If you attempt to create or update a user with a duplicate value for a custom user property with a uniqueness restriction, the user creation or update operation fails. The user isn't created or updated until you enter a unique value. Similarly, you need to provide a unique value for a unique custom group property to create or update a group.
```

Reviewer: I would add:
If the property is not also specified as being required, you can also omit the value entirely; multiple users or groups can omit the property and not violate uniqueness.

Author: Added:
```
`null` value doesn't come into the uniqueness calculation. If the unique property is not also specified as being required, you can also omit the value entirely; multiple users or groups can omit the property and not violate uniqueness.
```

> To enforce uniqueness for custom properties, you can either add new unique custom properties, or update existing custom properties to be unique.

Reviewer 2: ```suggestion
To enforce uniqueness for custom properties, you can either add new unique custom properties or update existing custom properties to be unique.
```

Reviewer: 👍

### Add new unique custom property

> You can use the [add property to user profile schema](#add-property-to-user-profile-schema) request or the [add property to group profile schema](#add-property-to-group-profile-schema) request to add one or more unique custom user or group properties, and specify `"unique": true` on the properties that need to be marked as unique.

Reviewer 2: ```suggestion
You can use the [add property to user profile schema](#add-property-to-user-profile-schema) request or the [add property to group profile schema](#add-property-to-group-profile-schema) request to add one or more unique custom user or group properties. Specify `"unique": true` on the properties that need to be marked as unique.
```

> ##### Request example

Author: ![Screen Shot 2021-06-30 at 12 24 51 PM](https://user-images.githubusercontent.com/12566503/124019964-8dae4280-d99e-11eb-890e-c15f5146c8e8.png)

                    "maxLength": 50,
                    "permissions": [
                        {
                            "principal": "SELF",
                            "action": "READ_WRITE"
                        }
                    ]
                },
                "email": {
                    "title": "Primary email",
                    "type": "string",
                    "required": true,
                    "format": "email",
                    "permissions": [
                        {
                            "principal": "SELF",
                            "action": "READ_WRITE"
                        }
                    ]
                }
            },
            "required": [
                "login",
                "firstName",
                "lastName",
                "email"
            ]
        },
        "custom": {
            "id": "#custom",
            "type": "object",
            "properties": {
              "twitterUserName": {
                  "title": "Twitter username",
                  "description": "User's username for twitter.com",
                  "type": "string",
                  "required": false,

>                   "unique": "UNIQUE_VALIDATED",

Reviewer: Do we explain anywhere how requesting `"unique": true` results in this?

Author: Added explanation at L2486.

                  "minLength": 1,
                  "maxLength": 20,
                  "permissions": [
                      {
                          "principal": "SELF",
                          "action": "READ_WRITE"
                      }
                  ]
              }
            },
            "required": []
        }
    },
    "type": "object",
    "properties": {
        "profile": {
            "allOf": [
                {
                    "$ref": "#/definitions/base"
                },
                {
                    "$ref": "#/definitions/custom"
                }
            ]
        }
    }
}
```

### Update existing custom property to be unique

> You can use the [update user profile schema property](#update-user-profile-schema-property) request or the [update group profile schema property](#update-group-profile-schema-property) request to mark existing custom user or group properties as unique, by specifying `"unique": true` on the properties that need to be marked as unique.

Reviewer 2: ```suggestion
You can use the [update user profile schema property](#update-user-profile-schema-property) request or the [update group profile schema property](#update-group-profile-schema-property) request to mark existing custom user or group properties as unique by specifying `"unique": true` on the properties that need to be marked as unique.
```

> Once the request to mark existing custom properties as unique is submitted, an asynchronous validation check is performed to make sure that there are no existing duplicate entries. If you have significant user or group records, the validation can take some time.

Reviewer 2: ```suggestion
After the request to mark existing custom properties as unique is submitted, an asynchronous validation check is performed to make sure that there are no existing duplicate entries. If you have significant user or group records, the validation can take some time.
```

Reviewer: "significant user or group records" --> "a significant number of users or groups"
(Let's not get into what would make an individual user or group record be significant.) :trollface:

> A uniqueness status `"unique": "PENDING_UNIQUENESS"` indicates that the validation check is still in progress. You can go to the Universal Directory UI page to track the status of the validation check. Once the validation completes, if you submit a [get user schema](#get-user-schema) request or a [get group schema](#get-group-schema) request the property's uniqueness status will be changed to `UNIQUE_VALIDATED` if no duplicate records are found and the property will be uniqueness enforced. Otherwise if duplicate records are found, the `unique` attribute of the schema property will not be shown in the get schema request and uniqueness will not be enforced on the schema property.

Reviewer 2: ```suggestion
A uniqueness status of `"unique": "PENDING_UNIQUENESS"` indicates that the validation check is still in progress. Use the Universal Directory page in the Admin Console (**Directory** > **Directory Integrations**) to track the status of the validation check. After the validation completes, if you submit a [get user schema](#get-user-schema) request or a [get group schema](#get-group-schema) request, the property's uniqueness status changes to `UNIQUE_VALIDATED` if no duplicate records are found. The property is then uniqueness enforced. Otherwise, if duplicate records are found, the `unique` attribute of the schema property isn't shown in the get schema request and uniqueness isn't enforced on the schema property.
```

Reviewer: Building on Susan's version: "The property is then uniqueness enforced." --> ", and uniqueness is then enforced on that property."

##### Request example

> For the request example below, assume that the default user profile schema contains a not unique custom property `twitterUserName`.

Author: ![Screen Shot 2021-06-30 at 12 25 02 PM](https://user-images.githubusercontent.com/12566503/124020072-adde0180-d99e-11eb-8f7a-89f775709871.png)

> For the request example below, assume that the default user profile schema contains a not unique custom property `twitterUserName`.

Reviewer 2: ```suggestion
For the request example below, assume that the default user profile schema contains the custom property `twitterUserName` that is not unique.
```

| `number`      | [JSON Number](https://tools.ietf.org/html/rfc7159#section-6) with double-precision 64-bit IEEE 754 floating point number constraint | `minimum` and `maximum`     |
| `integer`     | [JSON Number](https://tools.ietf.org/html/rfc7159#section-6) with 32-bit signed two's complement integer constraint           | `minimum` and `maximum`     |
| `array`       | [JSON Array](https://tools.ietf.org/html/rfc7159#section-5)                                                                         |                             |

> ## Unique attribute

Author: I wasn't sure whether it should be plural or singular. I thought it's the `unique` attribute of schema properties 🤔 . Changed to plural.

<ApiLifecycle access="ea" />

> You can enforce uniqueness for custom properties in the Okta user profile or the Okta group profile, such as an employee identification number. You can declare a maximum of five unique properties for each user type and five unique properties in the Okta group profile. Different user types can have the same or different unique properties (up to the limit of five per type).

Reviewer 3: "the Okta user profile" -> "Okta user profile(s)".  Let's not have _our own documentation_ ignore multiple user types :)

> Unique properties in the Okta user profile share a single namespace across all [user types](/docs/reference/api/user-types) in an org. If user types A and B both contain the property `ice cream` and you identify it as unique in both profiles, then if user type A has the value `chocolate`, no other users of type A or B (or any other user type with `ice cream` declared unique) can have that value.

Reviewer 3: See above

Reviewer 3: "if user type A" -> "if user U of type A"

Reviewer 3: Nit: `with 'ice cream' declared unique` -> `which declares 'ice cream'`

> Properties that aren't unique aren't tracked for uniqueness. If the property `candy` is unique in type E and not unique in type F, and a user of type E has the value `caramel` for the property, then no other users of type E can have the value `caramel` for the property. But any number of users of type F can have the value `caramel`. Although `candy` is unique in E, it isn't unique in F, so the value for the property in users of type F doesn't matter.

Reviewer 3: How about this:

> Properties that aren't unique also aren't tracked for uniqueness. Suppose the property `candy` is unique in type E and not unique in type F. If a user of type E has the value `caramel` for the `candy` property, no other users of type E can also have the value `caramel`. But any number of users of type F can subsequently be assigned the value `caramel`. Furthermore, because `candy` is not unique in F, any values users of type F may have are not considered when enforcing uniqueness for users of type E.  No matter how many users of type F already have the value `cotton`, it may be assigned to a user of type E as long as no other such user already has that value.

> Properties that aren't unique aren't tracked for uniqueness. If the property `candy` is unique in type E and not unique in type F, and a user of type E has the value `caramel` for the property, then no other users of type E can have the value `caramel` for the property. But any number of users of type F can have the value `caramel`. Although `candy` is unique in E, it isn't unique in F, so the value for the property in users of type F doesn't matter.

Author: I think Dayton's version elaborates better 😇 . Updated.

If you attempt to create or update a user with a duplicate value for a custom user property with a uniqueness restriction, the user creation or update operation fails. The user isn't created or updated until you enter a unique value. Similarly, you need to provide a unique value for a unique custom group property to create or update a group.

`null` value doesn't come into the uniqueness calculation. If the unique property is not also specified as being required, you can also omit the value entirely; multiple users or groups can omit the property and not violate uniqueness.

To enforce uniqueness for custom properties, you can either add new unique custom properties or update existing custom properties to be unique.

### Add new unique custom property

> You can use the [add property to user profile schema](#add-property-to-user-profile-schema) request or the [add property to group profile schema](#add-property-to-group-profile-schema) request to add one or more unique custom user or group properties. Specify `"unique": true` on the properties that need to be marked as unique. After the request is submitted, the property's uniqueness status is changed immediately to `"unique": "PENDING_UNIQUENESS"` immediately and uniqueness is then enforced on that property.

Reviewer: This is wrong.  For a new property it goes immediately to UNIQUE_VALIDATED.  And would it be correct to say the status is changed "when" the request is submitted, not "after"?  After all, the response to the request already shows it as validated, right?  So (note I also edited out the words "that need"):

Specify `"unique": true` on the properties to be marked as unique. The response shows the property with `"unique": "UNIQUE_VALIDATED"` and uniqueness is then enforced on that property.

Author: oops sorry the status was a copy-paste mistake.

Author: Should the later part of the sentence use plural since the first half mentions `properties`? Updated to:
```
Specify `"unique": true` on the properties to be marked as unique. The response shows the properties with `"unique": "UNIQUE_VALIDATED"` and uniqueness is then enforced on those properties.
```

                "login",
                "firstName",
                "lastName",
                "email"
            ]
        },
        "custom": {
            "id": "#custom",
            "type": "object",
            "properties": {
              "twitterUserName": {
                  "title": "Twitter username",
                  "description": "User's username for twitter.com",
                  "type": "string",
                  "required": false,
                  "unique": "UNIQUE_VALIDATED",
                  "minLength": 1,
                  "maxLength": 20,
                  "permissions": [
                      {
                          "principal": "SELF",
                          "action": "READ_WRITE"
                      }
                  ]
              }
            },
            "required": []
        }
    },
    "type": "object",
    "properties": {
        "profile": {
            "allOf": [
                {
                    "$ref": "#/definitions/base"
                },
                {
                    "$ref": "#/definitions/custom"
                }
            ]
        }
    }
}
```

### Update existing custom property to be unique

> You can use the [update user profile schema property](#update-user-profile-schema-property) request or the [update group profile schema property](#update-group-profile-schema-property) request to mark existing custom user or group properties as unique by specifying `"unique": true` on the properties that need to be marked as unique.

Reviewer: Again I suggest removing "that need", i.e. "... on the properties to be marked as unique."

            },
            "required": [
                "login",
                "firstName",
                "lastName",
                "email"
            ]
        },
        "custom": {
            "id": "#custom",
            "type": "object",
            "properties": {
              "twitterUserName": {
                  "title": "Twitter username",
                  "description": "User's username for twitter.com",
                  "type": "string",
                  "required": false,
                  "unique": "PENDING_UNIQUENESS",
                  "minLength": 1,
                  "maxLength": 10,
                  "permissions": [
                      {
                          "principal": "SELF",
                          "action": "READ_ONLY"
                      }
                  ]
              }
            },
            "required": []
        }
    },
    "type": "object",
    "properties": {
        "profile": {
            "allOf": [
                {
                    "$ref": "#/definitions/base"
                },
                {
                    "$ref": "#/definitions/custom"
                }
            ]
        }
    }
}

> ```

Reviewer: One big gap: You never show how to change a unique property to be non-unique.

Reviewer 3: Good point.  This is doubly worth calling out because the reverse operation (marking unique) is asynchronous, whereas I don't believe removing uniqueness is.

Author: Removing uniqueness is also "partially" asynchronous. The property becomes not unique immediately but a job is scheduled to cleanup the DB table. So it appears synchronous to the users.

Reviewer: Hm, I don't see the schema id in the `UniqueEntityProperty` table, just schema type (so that one user and one group can each have the same value for a like-named unique property).  So when property is changed from unique to non-unique in user type A, how do we manage to erase only those entries associated with users of that type?  (Looking... yes, I see the job does pass the type id into the scrubber method, and we do a Join with the user table to identify records for users of that type.)

So I was able to answer my own question there, but I still have this question: If a property is changed from unique to non-unique (scrubber job is scheduled) and then changed back to unique (validation job is scheduled), how do we handle the race between those two jobs?

| `number`      | [JSON Number](https://tools.ietf.org/html/rfc7159#section-6) with double-precision 64-bit IEEE 754 floating point number constraint | `minimum` and `maximum`     |
| `integer`     | [JSON Number](https://tools.ietf.org/html/rfc7159#section-6) with 32-bit signed two's complement integer constraint           | `minimum` and `maximum`     |
| `array`       | [JSON Array](https://tools.ietf.org/html/rfc7159#section-5)                                                                         |                             |

## Unique attributes

<ApiLifecycle access="ea" />

You can enforce uniqueness for custom properties in Okta user profiles or the Okta group profile, such as an employee identification number. You can declare a maximum of five unique properties for each user type and five unique properties in the Okta group profile. Different user types can have the same or different unique properties (up to the limit of five per type).

> Unique properties in Okta user profiles share a single namespace across all [user types](/docs/reference/api/user-types) in an org. If user types A and B both contain the property `ice cream` and you identify it as unique in both profiles, then if user U of type A has the value `chocolate`, no other users of type A or B (or any other user type which declares `ice cream` as unique) can have that value.

Reviewer: I agree with Dayton's earlier comment that the sentence should refer to a user and not a type as having the value, but I see no purpose to identifying the user as U since the identifier never appears again.  Why not:  ... if a user of type A has ...

> Properties that aren't unique also aren't tracked for uniqueness. Suppose the property `candy` is unique in type E and not unique in type F. If a user of type E has the value `caramel` for the `candy` property, no other users of type E can also have the value `caramel`. But any number of users of type F can subsequently be assigned the value `caramel`. Furthermore, because `candy` is not unique in type F, any values users of type F may have are not considered when enforcing uniqueness for users of type E. No matter how many users of type F already have the value `cotton`, it may be assigned to a user of type E as long as no other such user already has that value.

Reviewer: The word "subsequently" bugs me; it makes it sound as though users of type F cannot _already_ have the value `caramel`.  Also, we're back to having a sentence start with "But".  I suppose it's "old school" to be bothered by that, but I do feel those two sentences can afford to be combined.  Putting those complaints together:

If a user of type E has the value `caramel` for the `candy` property, no other users of type E can also have the value `caramel`, but any number of users of type F can already have or later be assigned the value `caramel`.

                  "minLength": 1,
                  "maxLength": 10,
                  "permissions": [
                      {
                          "principal": "SELF",
                          "action": "READ_ONLY"
                      }
                  ]
              }
            },
            "required": []
        }
    },
    "type": "object",
    "properties": {
        "profile": {
            "allOf": [
                {
                    "$ref": "#/definitions/base"
                },
                {
                    "$ref": "#/definitions/custom"
                }
            ]
        }
    }
}
```

### Update existing unique custom property to be non-unique

You can use the [update user profile schema property](#update-user-profile-schema-property) request or the [update group profile schema property](#update-group-profile-schema-property) request to change existing unique custom user or group properties to be non-unique by specifying `"unique": false` on the properties to be changed to non-unique. The response shows the properties without the `unique` attribute and the uniqueness constraint is then removed on those properties.

> Note that if multiple user types declare a property as unique and you removed the uniqueness constraint on one type, there may be a delay before users of other types which declare the property as unique can be assigned values formerly held by users of the first type.

Reviewer: tiny nit: `remove` instead of `removed`
